PYIC-7761: Recreate http client with tracing

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Recreate http client with tracing

### Why did it change

We introduced logic to recreate the base client in our tracing http client every hour. The new base client wasn't being created with instrumentation though, and so we lost things like spans in Dynatrace. This updates to recreate the client in the same way we create in initially.


### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7761](https://govukverify.atlassian.net/browse/PYIC-7761)


[PYIC-7761]: https://govukverify.atlassian.net/browse/PYIC-7761?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7761: Recreate http client with tracing

We introduced logic to recreate the base client in our tracing http
client every hour. The new base client wasn't being created with
instrumentation though, and so we lost things like spans in Dynatrace.
This updates to recreate the client in the same way we create in
initially.

Wynndow commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/tracing/TracingHttpClient.java:
```diff
@@ -108,6 +136,15 @@ public Optional<Executor> executor() {
+    private synchronized void recreateBaseClientIfNecessary() {
+        if (creationTime.plus(MAX_CLIENT_AGE).isBefore(Instant.now())) {
+            LOGGER.info("Recreating base HTTP client due to age");
+            this.baseClient = getInstrumetedHttpClient();
```
This line is the actual functional change here - most of the rest of the diff is just moving public methods above private ones.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/tracing/TracingHttpClient.java:
```diff
@@ -177,4 +174,10 @@ private static void addRequestInformation(Subsegment subsegment, HttpRequest req
         subsegment.putHttp("request", requestInformation);
     }
+
+    private static HttpClient getInstrumetedHttpClient() {
```
Typo `instrumented`

Wynndow replied:
Bloomin' words...

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/tracing/TracingHttpClient.java:
```diff
@@ -48,21 +48,49 @@ private TracingHttpClient(HttpClient baseClient) {
+            // for a long time (>1 hour). Retry them once, to recreate the connection.
+            // In the build environment we see connection resets for idle connections in
+            // the pool. Retrying uses a different connection.
+            if (HTTP_ERROR_MESSAGES_TO_RETRY.contains(e.getMessage())) {
```
While you're here... one thing I spotted was that your change to include Connection Reset changed this from a substring match to an exact match, which means the GOAWAYs never actually get retried.

I'm not sure retrying helps very much (and recreating the client seems to pretty much eliminate the issue), but ideally we should either make the logic work or remove it.

(Doesn't have to be part of this PR)

Wynndow replied:
Interesting. I'll update it now and pull out the retry stuff.

Wynndow committed:
PYIC-7761: Remove GOAWAY error handling

This problem is solved by the recreation of the client hourly, so no
need to check check it. The check was broken anyway.

Also shifts the client test file into a package to mirror the packages
in src.

Also fixes a typo.
